### PR TITLE
Read columns values by rowid and convert to write columns.

### LINKED ISSUE #2267: [Partial Update] read columns values by rowid
Implement TabletUpdates::_get_column_values, a subtask for https://github.com/StarRocks/starrocks/pull/2220

### PR DESCRIPTION
**Pull request by Linkerist**
I need to read column values and convert them to write columns.
Get rowids from a primary key column by primary-index, and then do plan_read_by_rssid.
After, read the corresponding segment file by rssid, and fetch value by rowids.

### CODE REVIEW

File: be/src/storage/tablet_updates.cpp [hunk 1]
```diff
@@ -2370,10 +2372,60 @@ void TabletUpdates::_update_total_stats(const std::vector<uint32_t>& rowsets) {
...
+        for (const auto& rowset : _rowsets) {
+            rssid_to_rowsets.insert(rowset);
+        }
+    }
+    for (const auto& [rssid, rowids] : rowids_by_rssid) {
+        auto iter = rssid_to_rowsets.upper_bound(rssid);
+        --iter;
+        const auto& rowset = iter->second.get();
+        CHECK_LE(rowset->rowset_meta()->get_rowset_seg_id(), rssid);
+        CHECK_LT(rssid, rowset->rowset_meta()->get_rowset_seg_id() + rowset->num_segments());
```
Comment (decster) [inline be/src/storage/tablet_updates.cpp:2390]: Do not crash BE, log error and use _set_error()

File: be/src/storage/tablet_updates.cpp [hunk 2]
```diff
@@ -2370,10 +2372,60 @@ void TabletUpdates::_update_total_stats(const std::vector<uint32_t>& rowsets) {
...
+        }
+        auto* seg_iter = down_cast<vectorized::SegmentIterator*>(iter_res.value().get());
+        if (seg_iter == nullptr) {
+            continue;
+        }
+        int i = 0;
+        for (const auto column_id : column_ids) {
+            if (with_default) {
+                (*columns)[i]->append_default();
+            }
```
Comment (decster) [inline be/src/storage/tablet_updates.cpp:2423]: Add default per segment?

File: be/test/storage/tablet_updates_test.cpp [hunk 3]
```diff
@@ -1477,4 +1477,36 @@ TEST_F(TabletUpdatesTest, snapshot_with_empty_rowset) {
...
+    _tablet = create_tablet(rand(), rand());
+    const int N = 8000;
+    std::vector<int64_t> keys;
+    for (int i = 0; i < N; i++) {
+        keys.push_back(i);
+    }
+    ASSERT_TRUE(_tablet->rowset_commit(2, create_rowset(_tablet, keys)).ok());
+    ASSERT_TRUE(_tablet->rowset_commit(3, create_rowset(_tablet, keys)).ok());
+    std::vector<uint32_t> read_column_ids = {1, 2};
+    std::map<uint32_t, std::vector<uint32_t>> rowids_by_rssid{{0, std::vector<uint32_t>{1, 2}}};
```
Comment (decster) [inline be/test/storage/tablet_updates_test.cpp:1491]: This rowids_by_rssid is too simple, so it doesn't find the earlier bug.

File: be/src/storage/tablet_updates.cpp [hunk 4]
```diff
@@ -2370,10 +2372,69 @@ void TabletUpdates::_update_total_stats(const std::vector<uint32_t>& rowsets) {
...
+            return iter_res.status();
+        }
+        auto* seg_iter = down_cast<vectorized::SegmentIterator*>(iter_res.value().get());
+        if (seg_iter == nullptr) {
+            continue;
+        }
+        int i = 0;
+        for (const auto column_id : column_ids) {
+            auto* col_iter = seg_iter->get_column_iterator(column_id);
+            col_iter->fetch_values_by_rowid(rowids.data(), rowids.size(), (*columns)[i].get());
```
Comment (decster) [inline be/src/storage/tablet_updates.cpp:2434]: check returned Status

**Review by decster — DISMISSED** [review]

File: be/src/storage/rowset_update_state.cpp [hunk 5]
```diff
@@ -208,6 +205,13 @@ Status RowsetUpdateState::apply(Tablet* tablet, Rowset* rowset, uint32_t rowset_
     }
     vector<std::unique_ptr<vectorized::Column>> read_columns(read_column_ids.size());
     vector<std::unique_ptr<vectorized::Column>> write_columns(read_column_ids.size());
+    for (auto i = 0; read_column_ids.size(); i++) {
```
Comment (DeepThinker666) [inline be/src/storage/rowset_update_state.cpp:208]:
```suggestion
    for (auto i = 0; i < read_column_ids.size(); i++) {
```

Comment (DeepThinker666) [reply be/src/storage/rowset_update_state.cpp:208]: better add an ut for this function

**Review by chaoyli — DISMISSED** [review]

File: be/src/storage/tablet_updates.cpp [hunk 6]
```diff
@@ -2370,10 +2372,69 @@ void TabletUpdates::_update_total_stats(const std::vector<uint32_t>& rowsets) {
...
+            column->append_default();
+        }
+    }
+    for (const auto& [rssid, rowids] : rowids_by_rssid) {
+        auto iter = rssid_to_rowsets.upper_bound(rssid);
+        --iter;
+        const auto& rowset = iter->second.get();
+        if (!(rowset->rowset_meta()->get_rowset_seg_id() <= rssid &&
+              rssid < rowset->rowset_meta()->get_rowset_seg_id() + rowset->num_segments())) {
+            string msg = Substitute("illegal rssid: $0, should in [$1, $2)", rssid,
```
Comment (DeepThinker666) [inline be/src/storage/tablet_updates.cpp:2396]:
```suggestion
            std::string msg = Substitute("illegal rssid: $0, should in [$1, $2)", rssid,
```

File: be/src/storage/tablet_updates.cpp [hunk 7]
```diff
@@ -2370,10 +2372,69 @@ void TabletUpdates::_update_total_stats(const std::vector<uint32_t>& rowsets) {
...
+            LOG(WARNING) << "Fail to open " << seg_path << ": " << segment_res.status();
+            return segment_res.status();
+        }
+        if ((*segment_res)->num_rows() == 0) {
+            continue;
+        }
+        vectorized::SegmentReadOptions seg_options;
+        seg_options.block_mgr = fs::fs_util::block_manager();
+        OlapReaderStatistics stats;
+        seg_options.stats = &stats;
```
Comment (DeepThinker666) [inline be/src/storage/tablet_updates.cpp:2418]: should use the stats of TabletReader

**Review by DeepThinker666 — CHANGES_REQUESTED** [review]

File: be/src/storage/tablet_updates.cpp [hunk 8]
```diff
@@ -2370,10 +2372,69 @@ void TabletUpdates::_update_total_stats(const std::vector<uint32_t>& rowsets) {
...
+            continue;
+        }
+        if (!iter_res.ok()) {
+            return iter_res.status();
+        }
+        auto* seg_iter = down_cast<vectorized::SegmentIterator*>(iter_res.value().get());
+        if (seg_iter == nullptr) {
+            continue;
+        }
+        int i = 0;
```
Comment (chaoyli) [inline be/src/storage/tablet_updates.cpp:2431]:
```suggestion
for (int i = 0; i < column_ids.size(); ++i) {
            auto* col_iter = seg_iter->get_column_iterator(column_ids[I]);
            RETURN_IF_ERROR(col_iter->fetch_values_by_rowid(rowids.data(), rowids.size() ,(*columns)[i].get()));
}
```

File: be/src/storage/tablet_updates.cpp [hunk 9]
```diff
@@ -2370,10 +2372,69 @@ void TabletUpdates::_update_total_stats(const std::vector<uint32_t>& rowsets) {
...
+            string msg = Substitute("illegal rssid: $0, should in [$1, $2)", rssid,
+                                    rowset->rowset_meta()->get_rowset_seg_id(),
+                                    rowset->rowset_meta()->get_rowset_seg_id() + rowset->num_segments());
+            LOG(ERROR) << msg;
+            _set_error();
+            return Status::InternalError(msg);
+        }
+        std::string seg_path =
+                BetaRowset::segment_file_path(rowset->rowset_path(), rowset->rowset_id(), rssid - iter->first);
+        auto segment_res =
```
Comment (chaoyli) [inline be/src/storage/tablet_updates.cpp:2405]:
```suggestion
        auto segment =
```

File: be/src/storage/tablet_updates.cpp [hunk 10]
```diff
@@ -2370,10 +2372,69 @@ void TabletUpdates::_update_total_stats(const std::vector<uint32_t>& rowsets) {
...
+                Segment::open(ExecEnv::GetInstance()->tablet_meta_mem_tracker(), fs::fs_util::block_manager(), seg_path,
+                              rssid - iter->first, &rowset->schema());
+        if (!segment_res.ok()) {
+            LOG(WARNING) << "Fail to open " << seg_path << ": " << segment_res.status();
+            return segment_res.status();
+        }
+        if ((*segment_res)->num_rows() == 0) {
```
Comment (chaoyli) [inline be/src/storage/tablet_updates.cpp:2412]: Why use *segment_res, not segment_res directly？

Comment (chaoyli) [reply be/src/storage/tablet_updates.cpp:2412]: Why num_rows() == 0 but not is_end_of_file()?

Comment (Linkerist) [reply be/src/storage/tablet_updates.cpp:2412]: > segment_res

the return value is statusor, not value directly, *segment_res is the value, segment_res is the status value

Comment (Linkerist) [reply be/src/storage/tablet_updates.cpp:2412]: > 

check segment num_row firstly, and then for latter segment iterator, use is_end_of_file.

File: be/src/storage/tablet_updates.cpp (hunk 9 above)
Comment (Linkerist) [reply be/src/storage/tablet_updates.cpp:2405]: the return value is statusor, not value directly, i try to point out this. but use 'segment' directly is also fine, changed.

File: be/src/storage/rowset/vectorized/segment_iterator.h [hunk 11]
```diff
@@ -5,15 +5,214 @@
...
 
 namespace starrocks::vectorized {
 
-class ColumnPredicate;
-class Schema;
-class SegmentReadOptions;
+/// SegmentIterator
+// TODO(zhuming): Refine the implementation of this class to reduce the intellectual overhead.
+// Too many policies encapsulated in this class, should split this class into many small classes.
+class SegmentIterator final : public ChunkIterator {
```
Comment (sduzh) [inline be/src/storage/rowset/vectorized/segment_iterator.h:23]: Why move this to the header file?

File: be/src/storage/tablet_updates.cpp (hunk 7 above)
Comment (Linkerist) [reply be/src/storage/tablet_updates.cpp]: write process, no stats dispatched.

File: be/src/storage/rowset/vectorized/segment_iterator.cpp [hunk 12]
```diff
@@ -69,201 +67,6 @@ static int compare(const SeekTuple& tuple, const Chunk& chunk) {
     return 0;
 }
 
-/// SegmentIterator
-// TODO(zhuming): Refine the implementation of this class to reduce the intellectual overhead.
-// Too many policies encapsulated in this class, should split this class into many small classes.
-class SegmentIterator final : public ChunkIterator {
```
[the remark this answers is not in the document]
Comment (Linkerist) [reply be/src/storage/rowset/vectorized/segment_iterator.cpp]: use column_iterator directly, instead segment iterator.

File: be/src/storage/rowset/vectorized/segment_iterator.h (hunk 11 above)
Comment (Linkerist) [reply be/src/storage/rowset/vectorized/segment_iterator.h]: use column_iterator directly, instead segment iterator.

File: be/src/storage/tablet_updates.cpp [hunk 13]
```diff
@@ -2370,10 +2376,71 @@ void TabletUpdates::_update_total_stats(const std::vector<uint32_t>& rowsets) {
...
+            const TabletColumn& tablet_column = _tablet.tablet_schema().column(column_ids[i]);
+            if (tablet_column.has_default_value()) {
+                const TypeInfoPtr& type_info = get_type_info(tablet_column);
+                std::unique_ptr<DefaultValueColumnIterator> default_value_iter =
+                        std::make_unique<DefaultValueColumnIterator>(
+                                tablet_column.has_default_value(), tablet_column.default_value(),
+                                tablet_column.is_nullable(), type_info, tablet_column.length(), 1);
+                ColumnIteratorOptions iter_opts;
+                RETURN_IF_ERROR(default_value_iter->init(iter_opts));
+                default_value_iter->fetch_values_by_rowid(nullptr, 1, (*columns)[i].get());
```
Comment (Linkerist) [inline be/src/storage/tablet_updates.cpp:2400]: Another way of implementation.
```cc
                auto* default_value = WrapperField::create(tablet_column);
                if (default_value == nullptr) {
                    return Status::InternalError("malloc error");
                }
                if (tablet_column.is_nullable() && tablet_column.default_value().length() == 0) {
                    default_value->set_null();
                } else {
                    default_value->from_string(tablet_column.default_value());
                }
                vectorized::Datum datum;
                if (default_value->is_null()) {
                    datum.set_null();
                } else {
                    std::string tmp = default_value->to_string();
                    Status st = datum_from_string(type_info.get(), &datum, tmp, nullptr);
                    if (!st.ok()) {
                        LOG(WARNING) << "create datum from string failed: status=" << st;
                        return st;
                    }
                }
                (*columns)[i]->append_datum(datum);
```

**Review by decster — APPROVED** [review]

**Review by chaoyli — APPROVED** [review]

**Comment by Linkerist** [comment]
resolve #2267

**Pull request merged**